### DISCUSSION TITLE
Empty component name when creating sbom from image created with bootBuildImage

### DISCUSSION THREAD

**Discussion by glennvdv** [aquasecurity/trivy#7556] [Bugs]
### Description

When creating an sbom of a docker image created with the command bootBuildImage. Trivy generates a component with an empty name.

### Desired Behavior

Or component should not be present, or name should be filled in

### Actual Behavior

name is missing

### Reproduction Steps

```bash
- generate spring boot project (https://start.spring.io/)
- run gradlew bootbuildImage
- run trivy image --format cyclonedx --output result.json IMAGE_NAME
Result file will contain something like this:
`    {
      "bom-ref": "f943ec78-bad5-4f04-99e7-2dcb106c6d85",
      "type": "application",
      "name": "",
      "properties": [
        {
          "name": "aquasecurity:trivy:Class",
          "value": "lang-pkgs"
        },
        {
          "name": "aquasecurity:trivy:Type",
          "value": "gobinary"
        }
      ]
    }`
```

### Target

Container Image

### Scanner

None

### Output Format

CycloneDX

### Mode

Standalone

### Debug Output

```bash
2024-09-19T13:36:50+02:00       DEBUG   No plugins loaded
2024-09-19T13:36:50+02:00       DEBUG   Default config file "file_path=trivy.yaml" not found, using built in values
2024-09-19T13:36:50+02:00       DEBUG   Cache dir       dir="/home/gv/.cache/trivy"
2024-09-19T13:36:50+02:00       DEBUG   Cache dir       dir="/home/gv/.cache/trivy"
2024-09-19T13:36:50+02:00       DEBUG   Parsed severities       severities=[UNKNOWN LOW MEDIUM HIGH CRITICAL]
2024-09-19T13:36:50+02:00       DEBUG   Ignore statuses statuses=[]
2024-09-19T13:36:50+02:00       INFO    "--format cyclonedx" disables security scanning. Specify "--scanners vuln" explicitly if you want to include vulnerabilities in the "cyclonedx" report.
2024-09-19T13:36:50+02:00       DEBUG   [pkg] Package types     types=[os library]
2024-09-19T13:36:50+02:00       DEBUG   [pkg] Package relationships     relationships=[unknown root direct indirect]
2024-09-19T13:36:50+02:00       DEBUG   Enabling misconfiguration scanners      scanners=[azure-arm cloudformation dockerfile helm kubernetes terraform terraformplan-json terraformplan-snapshot]
2024-09-19T13:36:50+02:00       DEBUG   Initializing scan cache...      type="fs"
2024-09-19T13:36:50+02:00       DEBUG   [image] Detected image ID       image_id="sha256:200e61309d3011bc0375d987b1610c61dcf87b0a59f4706defacaec5369db7d1"
2024-09-19T13:36:50+02:00       DEBUG   [image] Detected diff ID        diff_ids=[sha256:2573e0d8158209ed54ab25c87bcdcb00bd3d2539246960a3d592a1c599d70465 sha256:dfb662718104dfd7cf5b4894852d38b3dcc03ffd49e5c8b4a53e865b8154cdfd sha256:23fda1eb5bae018b1fb66118776fa7eb411610e591117db7929429460f65f7ac sha256:7cc16eb89e7d7c60c10030ea4f0299c03f8e28699fa54fbebb8c366741489213 sha256:343c6290b0f2ed03d2e417d111ed5db6e41914e338d83a45f73744d16899423f sha256:cb7c0abe7568a96bc898f79ddb15b400abb1c80f79f7b1361c2a9d183e53834f sha256:bea0a3dc2651cac7c9c567a5cb4e7536107b357cb9113e8806f690f050500012 sha256:3b37488d9a9c7539f69814901831aeef1c71777943180ffd1b14baf54345c4a4 sha256:417e5bfc3c82b9373cf6804206e071d2fc74560df867d0f39cb21ac3d15231b6 sha256:c6902f46e4f8c572b0ce611b53b2c8d0e5dcea7e3d35e41b8077619d31cef236 sha256:ce4e86e0307e01ceb95d3abacfa3e157fb781571d00e698faccf6fe91ccd83ee sha256:366ce7d1a7f90f2e4ad08752f87510eee3ffca18736fa63c03823c8c4ebf2925 sha256:8eb7d03527eed8a90ed088cc355e7c0e517118a31254c17d605f81862a92adf2 sha256:cca8781b24a9f6fad09dde178aa7f8559a5329e3c7dc19cc08f42ca47d0fca6d sha256:62b97a790014986d27ed578499435a06a269cc0a8894f3be4f32a8dd7806e2d2 sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef sha256:2257accaed4d6f1a6d11123af1ff5263787c0ade31f630a867976912a0ebdf94 sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef sha256:d877b693c335c5715d181ddb4c835af43cd48418f1edf4f7657eff8e24f8e4b9 sha256:f3e6835448298a6dcb929ee8a7f26c70b5693867cdc5807d4e3e5a1097d50b32 sha256:1dc94a70dbaa2171fb086500a5d27797f779219b126b0a1eebb9180c2792e80e]
2024-09-19T13:36:50+02:00       DEBUG   [image] Detected base layers    diff_ids=[]
2024-09-19T13:36:50+02:00       INFO    Detected OS     family="ubuntu" version="22.04"
2024-09-19T13:36:50+02:00       INFO    Number of language-specific files       num=6
2024-09-19T13:36:50+02:00       DEBUG   [vex] VEX filtering is disabled
```

### Operating System

linux

### Version

```bash
Version: 0.55.2
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-09-19 06:12:22.986098678 +0000 UTC
  NextUpdate: 2024-09-19 12:12:22.986098548 +0000 UTC
  DownloadedAt: 2024-09-19 06:41:53.411191991 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-09-17 01:05:13.063235569 +0000 UTC
  NextUpdate: 2024-09-20 01:05:13.063235449 +0000 UTC
  DownloadedAt: 2024-09-18 07:51:07.579646002 +0000 UTC
Check Bundle:
  Digest: sha256:ef2d9ad4fce0f933b20a662004d7e55bf200987c180e7f2cd531af631f408bb3
  DownloadedAt: 2024-09-18 09:28:51.515067159 +0000 UTC
```

### Checklist

- [ ] Run `trivy clean --all`
- [ ] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by DmitriyLewen** [comment]
Hello @glennvdv 

> Reproduction Steps

I couldn't build the image. It looks like the system needs some configuration.

We are focusing on other priorities.
It would be great if you could upload this test image to DockerHub.
It will help and speed up the investigation of this issue.

Regards, Dmitriy

**Comment by glennvdv** [comment]
What error are you getting? Only java & docker are needed to be able to build it

**Reply by DmitriyLewen** [reply]
```
...
    [creator]         Using Java version 23 extracted from MANIFEST.MF
    [creator]       No valid JRE available, providing matching JDK instead. Using a JDK at runtime has security implications.
    [creator]        : Contributing to layer
    [creator]       Warning: Dependency has no SHA256. Skipping cache.
    [creator]         Downloading from 
    [creator]     unable to invoke layer creator
    [creator]     unable to get dependency . see DEBUG log level
    [creator]     ERROR: failed to build: exit status 1

> Task :bootBuildImage FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':bootBuildImage'.
> Builder lifecycle 'creator' failed with status code 51

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

BUILD FAILED in 1m
5 actionable tasks: 5 executed

```

**Comment by glennvdv** [comment]
pushed to https://hub.docker.com/repository/docker/stuikerd/vulnerable-spring-boot/general
you can execute command 
`trivy image --format cyclonedx --output result.json  stuikerd/vulnerable-spring-boot:latest`
You will find following entry:
`    {
      "bom-ref": "8447fd4c-af4b-4d88-a392-84896a774682",
      "type": "application",
      "name": "",
      "properties": [
        {
          "name": "aquasecurity:trivy:Class",
          "value": "lang-pkgs"
        },
        {
          "name": "aquasecurity:trivy:Type",
          "value": "gobinary"
        }
      ]
    }`

**Reply by DmitriyLewen** [reply]
Hello @glennvdv 

I investigated this case.

This image contains `layers/sbom/launch/buildpacksio_lifecycle/launcher/sbom.spdx.json` file.

This file doesn't have application component  (SPDX format doesn't have strict restrictions in some points).

So Trivy detects only libraries, but can't correctly determinate filepath for application target.

**Reply by glennvdv** [reply]
Should this than be skipped or what should the correct behavior be?

**Reply by DmitriyLewen** [reply]
it is hard to say which behavior is correct.
The SPDX format doesn't have strict rules in some points, so we cannot foresee all possible cases.
But we always try to use the same structure for all Trivy reports (that is why we can't skip this component).
In this case, we cannot correctly determine which file path to use - so we leave this field empty.

**Comment by DmitriyLewen** [comment]
User found 1 more case with same problem - #8287.

This problem is aggravated by the fact that for such sbom files we create a spdx file with an empty name field (and this is not a valid file(#8287)

therefore we need to make some decision regarding such situations (fixes are needed in the sbom analyzer).

I see 2 ways:
- use a hash of the component structure or just uuid
- use the path to the sbom file. but this can cause confusion if we use the name of the sbom file and the name from the component for several components of one sbom)

@knqyf263 wdyt?

**Reply by knqyf263** [reply]
>use the path to the sbom file. but this can cause confusion if we use the name of the sbom file and the name from the component for several components of one sbom)

Do you mean if one sbom file contains several types of components, such as npm and python, then we will the sbom file path as the component name several times? I think it's better than empty, which makes our sbom invalid, and also, we can add a package type, somehow like `gobinary:layers/sbom/launch/buildpacksio_lifecycle/launcher/sbom.spdx.json`.

**Reply by DmitriyLewen** [reply]
I meant another case.

e.g., 
- one SBOM created with Trivy.
In this case, the SBOM contains an application component and we can determine the path to the application
e.g.:
```
➜ trivy -q image aquasec/trivy -o report.cdx.json -f cyclonedx 
➜ trivy -q sbom report.cdx.json -f cyclonedx | grep '"type": "application"' -A 1
          "type": "application",
          "group": "aquasecurity",
--
      "type": "application",
      "name": "usr/local/bin/trivy",
```

- and second SBOM from this discussion.

So Trivy will show different logic for these 2 SBOM files:
- `usr/local/bin/trivy` (path to App)
- `gobinary:layers/sbom/launch/buildpacksio_lifecycle/launcher/sbom.spdx.json (path to SBOM)

**Reply by knqyf263** [reply]
Ideally, we shouldn't add an application component in that case, right?

**Reply by DmitriyLewen** [reply]
I don't think this is a solution either.
SBOM may contain a component of a different type.
for example:
```
{
      "name": "launcher",
      "SPDXID": "SPDXRef-DocumentRoot-File-launcher",
      "versionInfo": "sha256:716665ae98fb4b4675d5184f80884547597d47be1395d1049dc9e16035f32cc1",
      "supplier": "NOASSERTION",
      "downloadLocation": "NOASSERTION",
      "filesAnalyzed": false,
      "checksums": [
        {
          "algorithm": "SHA256",
          "checksumValue": "716665ae98fb4b4675d5184f80884547597d47be1395d1049dc9e16035f32cc1"
        }
      ],
      "licenseConcluded": "NOASSERTION",
      "licenseDeclared": "NOASSERTION",
      "primaryPackagePurpose": "FILE"
    }
```

then we'll just lose this component, and that will be worse than confusion (and it's not a fact that it will confuse users) with file paths.

**Reply by DmitriyLewen** [reply]
I have been returning to this problem for a long time and came to the conclusion that we should still use 2 possible paths (path to the application and path to the SBOM file)

I will describe this situation in the documentation.
And if there are too many questions - I will "look for answers" on the best path to the file from users.

**Reply by knqyf263** [reply]
I meant we don't need to inject pseudo applications aggregating package components and should directly add package components if they don't have an application parent in SBOM. But I may be missing something. I don't understand [this example](https://github.com/aquasecurity/trivy/discussions/7556#discussioncomment-12051436). It just says `launcher`. How can we know the package type?

**Reply by DmitriyLewen** [reply]
> I meant we don't need to inject pseudo applications aggregating package components and should directly add package components if they don't have an application parent in SBOM

My first thought was - "yes. This should work. Let's do it".
But after thinking more, I realized that this could be a problem for users.
For example, an image contains 10 SBOM files (without the Application component). So we will add relationships between the image component and all direct packages.
As you can understand, it will be a heap with all the packages.
So it is better to add pseudo-applications with the path to the sbom files.

> I don't understand https://github.com/aquasecurity/trivy/discussions/7556#discussioncomment-12051436. It just says launcher. How can we know the package type?

I just made a simplified SBOM (for testing).
You can see both components, relationships, files here:
```json
{
  "spdxVersion": "SPDX-2.3",
  "dataLicense": "CC0-1.0",
  "SPDXID": "SPDXRef-DOCUMENT",
  "name": "launcher",
  "documentNamespace": "https://anchore.com/syft/file/launcher-268e7779-ba66-4422-a5b0-d4d83f7b5d8c",
  "creationInfo": {
    "licenseListVersion": "3.25",
    "creators": [
      "Organization: Anchore, Inc",
      "Tool: syft-1.13.0"
    ],
    "created": "2024-09-25T21:11:50Z"
  },
  "packages": [
    {
      "name": "github.com/buildpacks/lifecycle",
      "SPDXID": "SPDXRef-Package-go-module-github.com-buildpacks-lifecycle-89c3bd8d4c2e75b7",
      "versionInfo": "v0.20.2",
      "supplier": "NOASSERTION",
      "downloadLocation": "NOASSERTION",
      "filesAnalyzed": false,
      "sourceInfo": "acquired package info from go module information: /launcher",
      "licenseConcluded": "NOASSERTION",
      "licenseDeclared": "NOASSERTION",
      "copyrightText": "NOASSERTION",
      "externalRefs": [
        {
          "referenceCategory": "SECURITY",
          "referenceType": "cpe23Type",
          "referenceLocator": "cpe:2.3:a:buildpacks:lifecycle:v0.20.2:*:*:*:*:*:*:*"
        },
        {
          "referenceCategory": "PACKAGE-MANAGER",
          "referenceType": "purl",
          "referenceLocator": "pkg:golang/github.com/buildpacks/lifecycle@v0.20.2"
        }
      ]
    },
    {
      "name": "launcher",
      "SPDXID": "SPDXRef-DocumentRoot-File-launcher",
      "versionInfo": "sha256:716665ae98fb4b4675d5184f80884547597d47be1395d1049dc9e16035f32cc1",
      "supplier": "NOASSERTION",
      "downloadLocation": "NOASSERTION",
      "filesAnalyzed": false,
      "checksums": [
        {
          "algorithm": "SHA256",
          "checksumValue": "716665ae98fb4b4675d5184f80884547597d47be1395d1049dc9e16035f32cc1"
        }
      ],
      "licenseConcluded": "NOASSERTION",
      "licenseDeclared": "NOASSERTION",
      "primaryPackagePurpose": "FILE"
    }
  ],
  "files": [
    {
      "fileName": "/launcher",
      "SPDXID": "SPDXRef-File-launcher-361242815a383bec",
      "checksums": [
        {
          "algorithm": "SHA1",
          "checksumValue": "0000000000000000000000000000000000000000"
        }
      ],
      "licenseConcluded": "NOASSERTION",
      "licenseInfoInFiles": [
        "NOASSERTION"
      ],
      "copyrightText": ""
    }
  ],
  "relationships": [
    {
      "spdxElementId": "SPDXRef-Package-go-module-github.com-buildpacks-lifecycle-89c3bd8d4c2e75b7",
      "relatedSpdxElement": "SPDXRef-File-launcher-361242815a383bec",
      "relationshipType": "OTHER",
      "comment": "evident-by: indicates the package's existence is evident by the given file"
    },
    {
      "spdxElementId": "SPDXRef-DocumentRoot-File-launcher",
      "relatedSpdxElement": "SPDXRef-Package-go-module-github.com-buildpacks-lifecycle-89c3bd8d4c2e75b7",
      "relationshipType": "CONTAINS"
    },
    {
      "spdxElementId": "SPDXRef-DOCUMENT",
      "relatedSpdxElement": "SPDXRef-DocumentRoot-File-launcher",
      "relationshipType": "DESCRIBES"
    }
  ]
}
```

We can't detect type from `SPDXRef-DocumentRoot-File-launcher`.
We detect `gobinary` type from packages.

**Reply by knqyf263** [reply]
>As you can understand, it will be a heap with all the packages.

Yes, that's what the original SBOM describes. Shouldn't we keep it? Also, we can add `PackageSourceInfo` so that users won't be confused.

```
PackageSourceInfo: <text><package> from /usr/share/sbom/container.spdx.json</text>
```

**Reply by DmitriyLewen** [reply]
> Yes, that's what the original SBOM describes. Shouldn't we keep it?

This will not be the original SBOM.

Take a look at the SBOM from my example.
This SBOM shows Application, but Trivy can't "read" it.
So we need to add all the unreadable (for Trivy) components (e.,g. `SPDXRef-DocumentRoot-File-launcher`). Otherwise we will get a non-original SBOM.

Or am i missing something?

**Reply by knqyf263** [reply]
>This SBOM shows Application, but Trivy can't "read" it.

Ah, I get your point now. I was not following what you were talking about. The root cause is that Trivy cannot identify the application component from third-party SBOMs.
Anyway, is it problematic if we flatten the packages in the SBOMs? The application component we inject is pseudo. I'm not sure if it helps users understand. Then, isn't it simpler to flatten them and add `PackageSourceInfo` to clarify those packages come from SBOM?

**Reply by DmitriyLewen** [reply]
> PackageSourceInfo to clarify those packages come from SBOM?

Do you want to add `PackageSourceInfo` as property, right?

> The application component we inject is pseudo. I'm not sure if it helps users understand. Then, isn't it simpler to flatten them

Pseudo application is more logical for me. But perhaps I'm just accustomed to this logic.
 
One question I see from users we'll get is `why for the same types Trivy may or may not include an app component?`.


For me it's better to use the same structure for all cases.
It's also easier for users to understand why reports include different filePaths rather than different component chains (`root->app->pkg` vs `root->pkg`).


but like I said - that might just be my habit.

**Reply by knqyf263** [reply]
>Do you want to add PackageSourceInfo as property, right?

Sorry I was talking about SPDX. Doesn't CycloneDX have a similar field?

>Pseudo application is more logical for me.

OK, we anyway need a short-term fix since Trivy is currently generating invalid SBOMs. If injecting pseudo applications is the easiest way, it's good to go for it at the moment.

**Reply by DmitriyLewen** [reply]
> Doesn't CycloneDX have a similar field?

I don't see a suitable field. So I would use properties.

 > OK, we anyway need a short-term fix since Trivy is currently generating invalid SBOMs. If injecting pseudo applications is the easiest way, it's good to go for it at the moment.
 
#8346

**Comment by DmitriyLewen** [comment]
Related issue - #8189
